src/components/CytoscapeGraph/Layout/BoxLayout.ts

```diff
@@ -0,0 +1,265 @@
+    const sourceId = this.normalizeToParent(source).id();
+    const targetId = this.normalizeToParent(target).id();
+
+    if (sourceId === targetId) {
+      return false;
+    }
+
+    const key = `${sourceId}->${targetId}`;
+
+    if (this.generatedMap[key]) {
+      return false;
+    }
+
+    this.generatedMap[key] = true;
+
+    return {
+      group: 'edges',
+      data: {
+        id: 'synthetic-edge-' + this.nextId++,
+        source: sourceId,
+        target: targetId
+      }
+    };
+  }
+
+  // Returns the parent if any or the element itself.
+  private normalizeToParent(element: any) {
+    return element.isChild() ? element.parent() : element;
+  }
+}
+
+/**
+ * Main class for the BoxLayout, used to bridge with cytoscape to make it easier to integrate with current code
+ */
+export default class BoxLayout {
+  readonly options;
+  readonly cy;
+  readonly elements;
+  readonly syntheticEdgeGenerator;
+
+  constructor(options: any) {
+    this.options = { ...options };
+    this.cy = this.options.cy;
+    this.elements = this.options.eles;
+    this.syntheticEdgeGenerator = new SyntheticEdgeGenerator();
+  }
+
+  /**
+   * This code gets executed on the cy.layout(...).run() is our entrypoint of this algorithm.
+   */
+  run() {
+    const { appBoxLayout, clusterBoxLayout, defaultLayout, namespaceBoxLayout } = this.options;
+    let allBoxNodes = this.cy.collection();
+    let removedElements = this.cy.collection();
+    let syntheticEdges = this.cy.collection();
+
+    for (const boxByType of [BoxByType.APP, BoxByType.NAMESPACE, BoxByType.CLUSTER]) {
+      const boxNodes = this.getBoxNodes(boxByType);
+      allBoxNodes = allBoxNodes.add(boxNodes);
+
+      let boxLayoutOptions = defaultLayout;
+      switch (boxByType) {
+        case BoxByType.APP:
+          if (appBoxLayout) {
+            boxLayoutOptions = appBoxLayout;
+          }
+          break;
+        case BoxByType.CLUSTER:
+          if (clusterBoxLayout) {
+            boxLayoutOptions = clusterBoxLayout;
+          }
+          break;
+        case BoxByType.NAMESPACE:
+          if (namespaceBoxLayout) {
+            boxLayoutOptions = namespaceBoxLayout;
+          }
+          break;
+      }
+
+      // (1.a) Prepare box node by assigning a size and running the compound layout
+      boxNodes.each(boxNode => {
+        const boxedNodes = boxNode.children();
+        const boxedElements = boxedNodes.add(boxedNodes.edgesTo(boxedNodes));
+        const boxLayout = boxedElements.layout(boxLayoutOptions);
+
+        boxLayout.on('layoutstart layoutready layoutstop', _evt => {
+          // Avoid propagating any local layout events up to cy, this would yield a global operation before the nodes are ready.
+          return false;
+        });
+
+        // synch layouts (dagre) stop before run() returns, async layouts (cose,cola) don't, so
+        // wait for the async layouts to stop before continuing.
+        this.waitForLayout(boxLayout);
```

Comment:
Does this work? AFAIK it is not really waiting outside of the `waitForLayout`.
`waitForLayout` is an `async function`, meaning that it is really returning a promise which you need to `await` here so it "waits". The async will bubble up to the top and this layout would need to be an async layout too.

Reply:
No, it does not work :)  But I fixed it...